File: app/javascript/channels/chat_room_channel.js
import consumer from "./consumer"

const chatRoomId = window.chatRoomId

consumer.subscriptions.create(
  { channel: "ChatRoomChannel", id: chatRoomId },
  {
    connected() {
      console.log("✅ WebSocket connected")
    },

    disconnected() {
      console.log("❌ WebSocket disconnected")
    },

    received(data) {
      if (data.type === "typing") {
        document.getElementById("typing").innerText =
          `${data.user} is typing...`

        setTimeout(() => {
          document.getElementById("typing").innerText = ""
        }, 1500)
      } else {
        document.getElementById("messages")
          .insertAdjacentHTML("beforeend", data)
      }
    },

    typing() {
      this.perform("typing")
    }
  }
)

File: app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    @chat_room = ChatRoom.find(params[:id])
    stream_for @chat_room
  end

  def typing(data)
    ChatRoomChannel.broadcast_to(
      @chat_room,
      {
        type: "typing",
        user: current_user.email
      }
    )
  end
end
